rust/ql/lib/codeql/rust/elements/internal/LocatableImpl.qll

```diff
@@ -17,17 +18,39 @@ module Impl {
   class Locatable extends Generated::Locatable {
     pragma[nomagic]
     final Location getLocation() {
-      exists(Raw::Locatable raw |
-        raw = Synth::convertLocatableToRaw(this) and
-        (
-          locatable_locations(raw, result)
+      exists(Raw::Locatable raw | raw = Synth::convertLocatableToRaw(this) |
+        exists(@location_default location |
+          result = LocationImpl::TLocationDefault(location) and
+          locatable_locations(raw, location)
+        )
+        or
+        not locatable_locations(raw, _) and
+        exists(File file, int beginLine, int beginColumn, int endLine, int endColumn |
+          this.hasLocationInfo(file.getAbsolutePath(), beginLine, beginColumn, endLine, endColumn)
+        |
+          result = LocationImpl::TLocationSynth(file, beginLine, beginColumn, endLine, endColumn)
           or
-          not exists(Location loc | locatable_locations(raw, loc)) and
-          result instanceof EmptyLocation
+          exists(@location_default location |
+            result = LocationImpl::TLocationDefault(location) and
+            locations_default(location, file, beginLine, beginColumn, endLine, endColumn)
+          )
         )
       )
     }
 
+    /**
+     * Holds if this element is at the specified location.
+     * The location spans column `startcolumn` of line `startline` to
+     * column `endcolumn` of line `endline` in file `filepath`.
+     * For more information, see
+     * [Providing locations in CodeQL queries](https://codeql.github.com/docs/writing-codeql-queries/providing-locations-in-codeql-queries/).
+     */
+    predicate hasLocationInfo(
```

Comment:
I think it would be better to not have this, but I see why it is needed currently. Instead, I would suggest having

```ql
abstract class SynthLocatable extends Locatable {
  abstract predicate predicate hasSynthLocationInfo(string filepath, int startline, int startcolumn, int endline, int endcolumn);
}
```

and then inside `LocationImpl.qll` you can replace
```ql
      exists(string filePath | file.getAbsolutePath() = filePath |
        any(Format f).hasLocationInfo(filePath, beginLine, beginColumn, endLine, endColumn) or
        any(FormatArgument f).hasLocationInfo(filePath, beginLine, beginColumn, endLine, endColumn)
      )
```
with
```ql
      any(SynthLocatable l).hasSynthLocationInfo(filePath, beginLine, beginColumn, endLine, endColumn)
```

and you then need to add

```ql
private class FormatSynthLocationImpl extends Format, SynthLocatable {
  override predicate predicate hasSynthLocationInfo(string filepath, int startline, int startcolumn, int endline, int endcolumn) { ... }
}
```
and similar for `FormatArgument`.

The advantage is that we restrict it to actual synthetic locations, and we do not "pollute" the user-facing classes with the `has(Synth)LocationInfo` predicate.